### PR TITLE
chore: add property tests for ABI encoding

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/5226

## Summary\*

This PR adds property testing to ABI encoding so that we can show that any valid `Abi` + `InputMap` combination can be abi encoded and decoded to return the original inputs.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Have dug into proptest strategies a bit more and we can avoid building the type at the same time as the value with sboxed strategies. We can leave that to a refactoring in a different PR however.

File: tooling/noirc_abi/src/lib.rs [hunk 1]
```diff
@@ -243,16 +250,21 @@ impl AbiParameter {
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(test, derive(arbitrary::Arbitrary))]
+
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/lib.rs:254]:
```suggestion
```

File: tooling/noirc_abi/src/arbitrary.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        AbiType::Boolean => {
+            any::<bool>().prop_map(|val| InputValue::Field(FieldElement::from(val))).sboxed()
+        }
+
+        AbiType::String { length } => {
+            // Strings only allow ASCII characters as each character must be able to be represented by a single byte.
+            let string_regex = format!("[[:ascii:]]{{{length}}}");
+            proptest::string::string_regex(&string_regex)
+                .expect("fix me")
```
Comment (michaeljklein) [inline tooling/noirc_abi/src/arbitrary.rs:51]: leftover TODO? it looks like this will always succeed

File: tooling/noirc_abi/src/arbitrary.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        (1..MAX_STRING_LEN).prop_map(|length| AbiType::String { length }),
+    ]
+    .sboxed()
+}
+
+pub(super) fn arb_abi_type() -> BoxedStrategy<AbiType> {
+    let leaf = arb_primitive_abi_type();
+
+    leaf.prop_recursive(
+        8,   // 8 levels deep
```
Comment (michaeljklein) [inline tooling/noirc_abi/src/arbitrary.rs:102]: nit: `up to 8 levels deep`, had to check docs to see whether it'd ever be `0` levels deep (i.e. a primitive type)

**Review by michaeljklein — APPROVED** [review]
two nits that aren't blocking: LGTM

**Pull request merged**